server/log_operation_manager.go

```diff
@@ -96,8 +98,65 @@ func (l LogOperationManager) getLogsAndExecutePass(ctx context.Context) error {
 		return err
 	}
 
-	// Process each active log once.
-	l.logOperation.ExecutePass(ctx, logIDs, &l.info)
+	numWorkers := l.info.numWorkers
+	if numWorkers == 0 {
+		glog.Warning("Executing a LogOperation pass with numWorkers == 0, assuming 1")
+		numWorkers = 1
+	}
+	glog.V(1).Infof("Beginning run for %v active log(s) using %d workers", len(logIDs), numWorkers)
+
+	var mu sync.Mutex
+	successCount := 0
+	itemCount := 0
+
+	// Build a channel of the logIDs that need to be processed.
+	toProcess := make(chan int64, len(logIDs))
+	for _, logID := range logIDs {
+		toProcess <- logID
+	}
+	close(toProcess)
+
+	// Set off a collection of worker goroutines to process the pending logIDs.
+	startBatch := time.Now()
+	var wg sync.WaitGroup
+	for i := 0; i < numWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				logID, more := <-toProcess
+				if !more {
+					return
+				}
+
+				ctx = util.NewLogContext(ctx, logID)
+				start := time.Now()
+				count, err := l.logOperation.ExecutePass(ctx, logID, &l.info)
+
+				if err == nil {
+					if count > 0 {
+						d := time.Now().Sub(start).Seconds()
+						glog.Infof("%v: processed %d items in %.2f seconds (%.2f qps)", logID, count, d, float64(count)/d)
+					} else {
+						glog.V(1).Infof("%v: no items to process", logID)
+					}
+					mu.Lock()
+					successCount++
+					itemCount += count
+					mu.Unlock()
+				}
+			}
+		}()
+	}
+
+	// Wait for the workers to consume all of the logIDs
+	wg.Wait()
+	d := time.Now().Sub(startBatch).Seconds()
+
+	mu.Lock()
+	glog.V(1).Infof("Group run completed in %.2f seconds: %v succeeded, %v failed, %v items processed", d, successCount, len(logIDs)-successCount, itemCount)
```

Comment:
Not sure of the reason for the lock as the wait group is done. Keeping lock analysis tools happy?